src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4740,6 +4893,28 @@ void StorageReplicatedMergeTree::startupImpl(bool from_attach_thread)
 }
 
 
+void StorageReplicatedMergeTree::flushAndPrepareForShutdown()
+{
+    if (shutdown_prepared_called.exchange(true))
+        return;
+
+    auto settings_ptr = getSettings();
+    /// Cancel fetches, merges and mutations to force the queue_task to finish ASAP.
+    fetcher.blocker.cancelForever();
+    merger_mutator.merges_blocker.cancelForever();
+    parts_mover.moves_blocker.cancelForever();
+    mutations_finalizing_task->deactivate();
+    stopBeingLeader();
+
+    if (attach_thread)
+        attach_thread->shutdown();
+
+    restarting_thread.shutdown(/* part_of_full_shutdown */true);
```

Comment:
There's still a race condition between `mutations_finalizing_task->deactivate()` and `mutations_finalizing_task->activateAndSchedule()` in `ReplicatedMergeTreeRestartingThread::runImpl`